Core/UI/UIManager.cs

```diff
@@ -0,0 +1,283 @@
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Terraria.UI;
+
+namespace PathOfTerraria.Core.UI;
+
+/// <summary>
+///		Manages the registration, updating, and rendering of <see cref="UIState"/> instances.
+/// </summary>
+/// <remarks>
+///		Registered <see cref="UIState"/> instances are identified by a unique identifier, which determines
+///		their behavior and associated data. They will be automatically unloaded when no longer needed.
+///		<br />
+///		This system allows multiple entries of the same <see cref="UIState"/> type, provided each entry has
+///		a distinct identifier.
+/// </remarks>
+[Autoload(Side = ModSide.Client)]
+public sealed class UIManager : ModSystem
+{
+	private sealed class UIStateData<T> where T : UIState
+	{
+		/// <summary>
+		///		Whether the state is enabled or not.
+		/// </summary>
+		public bool Enabled;
+		
+		/// <summary>
+		///		The <see cref="UserInterface"/> instance associated with the <see cref="UIState"/>.
+		/// </summary>
+		public UserInterface UserInterface;
+		
+		/// <summary>
+		///		The identifier of the <see cref="UIState"/>.
+		/// </summary>
+		public readonly string Identifier;
+		
+		/// <summary>
+		///		The layer at which to insert the <see cref="UIState"/>.
+		/// </summary>
+		public readonly string Layer;
+
+		/// <summary>
+		///		The value of the <see cref="UIState"/>.
+		/// </summary>
+		public readonly T? Value;
+
+		/// <summary>
+		///		The index offset within the specified insertion layer. 
+		/// </summary>
+		/// <remarks>
+		///		Defaults to <c>0</c>.
+		/// </remarks>
+		public readonly int Offset;
+
+		/// <summary>
+		///		The interface scale type of the <see cref="UIState"/>. 
+		/// </summary>
+		/// <remarks>
+		///		Defaults to <see cref="InterfaceScaleType.UI"/>.
+		/// </remarks>
+		public readonly InterfaceScaleType Type;
+		
+		public UIStateData(string identifier, string layer, T? value, int offset = 0, InterfaceScaleType type = InterfaceScaleType.UI)
```

Comment:
Thoughts on using a primary ctor? Would shorten the file a bit and be just as readable. Fully opinion; we might want to discuss if we use primary ctors when possible or not.

Reply:
Oh I think it's purely yet another formatting design decision. The mod seems to make a wide use of it so far, so sure, why not.